include/mbedtls/aes.h

```diff
@@ -212,6 +281,72 @@ int mbedtls_aes_crypt_cbc( mbedtls_aes_context *ctx,
                     unsigned char *output );
 #endif /* MBEDTLS_CIPHER_MODE_CBC */
 
+#if defined(MBEDTLS_CIPHER_MODE_XTS)
+/**
+ * \brief      This function performs an AES-XEX encryption or decryption
+ *             operation for an entire XEX data unit.
+ *
+ *             Length must be a multiple of the AES block size (16 bytes).
+ *
+ * \param ctx          The AES XTS context to use for AES XEX operations.
+ * \param mode         The AES operation: #MBEDTLS_AES_ENCRYPT or
+ *                     #MBEDTLS_AES_DECRYPT.
+ * \param length       The of the input data in bytes. The length must be a
+ *                     multiple of 16 bytes.
+ * \param iv           The initialization vector (not updated after use).
+ * \param input        The buffer holding the input data.
+ * \param output       The buffer holding the output data.
+ *
+ * \return             \c 0 on success.
+ * \return             #MBEDTLS_ERR_AES_INVALID_INPUT_LENGTH if length smaller
+ *                     than an AES block in size (16 bytes) or if the length is
+ *                     larger than 2**20 blocks.
+ */
+int mbedtls_aes_crypt_xex( mbedtls_aes_xts_context *ctx,
```

Comment:
I am skeptical that we need an XEX entry point, since mechanically, it's just a special case of XTS. However, the policy checks we do up front are different. XTS has a limit on the number of blocks, for instance, but (despite what the incorrect comment says) there is no such limit on XEX.

That said, the tests we run are largely duplicate. XEX tests could work with our XTS entry point so long as the number of blocks the test tries to use is not too large (i.e. less than 2 ** 20 blocks, a restriction from NIST SP 800-38E).

Reply:
The reason for the suggested limit is probably that the bit security the security proof guaranties decreases as the size of the data unit rises and at 2**20 blocks the bit security reaches 85.

Typical sector sizes are 0.5kB, and 4kB and 2**20 blocks is 16MB. How likely is it that 16MB is not enough for the foreseeable future?

Reply:
I don't know how likely it is to be enough for the foreseeable future. It seems like it'd be enough, but another question is: do we want to enforce this policy in our library? XEX doesn't require it. Only the NIST XTS standard requires this.

Reply:
As I mentioned above the limit is a safe and reasonable choice. If you go above the limit you are not clearly broken, but your security will be below what is considered desirable today.

I've been thinking if we could enforce the limit for now and think it through again if and when somebody complains?

Reply:
I'll enforce the limit on XTS.

Are you proposing adding the same limit to the XEX function? How I've currently implemented it, the XEX function is just another entry point to the underlying XTS implementation, but with different policy checks. We could also consider removing the XEX function as the XTS entry point is good enough for all known use cases.

Reply:
Additionally, IEEE P1619 defines an XEX construction that uses two keys, unlike the Rogaway paper. So, there is disagreement about what is meant exactly by XEX. This is more reason to remove the XEX API, unless we decide to pick a standard/paper.

Reply:
I've removed XEX, since the XTS entry point should be good enough for the applications we know of and XEX isn't standardized.